src/iris/utils.py

```diff
@@ -197,26 +213,43 @@ def claim_incident(incident_id, owner, session=None):
 def claim_bulk_incidents(incident_ids, owner):
     session = db.Session()
     incident_ids = tuple(incident_ids)
-    active = 0 if owner else 1
     now = datetime.datetime.utcnow()
-    session.execute('''UPDATE `incident`
-                       SET `incident`.`updated` = :updated,
-                           `incident`.`active` = :active,
-                           `incident`.`owner_id` = (SELECT `target`.`id` FROM `target` WHERE `target`.`name` = :owner AND `type_id` = (SELECT `id` FROM `target_type` WHERE `name` = 'user'))
-                       WHERE `incident`.`id` IN :incident_ids''',
-                    {'incident_ids': incident_ids, 'active': active, 'owner': owner, 'updated': now})
 
-    session.execute('UPDATE `message` SET `active`=0 WHERE `incident_id` IN :incident_ids', {'incident_ids': incident_ids})
+    action = 'claim' if owner else 'unclaim'
+    for incident_id in incident_ids:
+        session.execute('''
+            INSERT INTO `incident_claim` (`incident_id`, `time`, `action_id`, `target_id`)
+            VALUES (
+                :incident_id,
+                :now,
+                (SELECT `incident_claim_action`.`id` FROM `incident_claim_action` WHERE `incident_claim_action`.`name` = :action),
+                (SELECT `target`.`id` FROM `target` WHERE `target`.`name` = :owner AND `type_id` = (SELECT `id` FROM `target_type` WHERE `name` = 'user'))
+            )
+        ''', {'incident_id': incident_id, 'owner': owner, 'now': now, 'action': action})
+
     session.commit()
 
+    session.execute('UPDATE `message` SET `active`=0 WHERE `incident_id` IN :incident_ids', {'incident_ids': incident_ids})
+
     claimed = set()
     not_claimed = set()
 
-    for incident_id, active in session.execute('SELECT `id`, `active` FROM `incident` WHERE `id` IN :incident_ids', {'incident_ids': incident_ids}):
-        if active == 1:
-            not_claimed.add(incident_id)
-        else:
+    check_claimed_sql = '''
+      SELECT `id`, (
```

Comment:
Could be made more efficient with joins, but if :incident_ids is small, it won't really matter since the dataset will be so small.

Reply:
Updated this to use join instead of subquery.